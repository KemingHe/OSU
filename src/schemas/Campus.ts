// ./src/schemas/Campus.ts
//
// OSU Campus ZodSchema and type definition,
// plus corresponding native TypeScript tuple definition.

// Zod essential imports.
import { type ZodSchema, z } from "zod";

export const CampusTuple = [
  "Columbus",
  "Lima",
  "Marion",
  "Mansfield",
  "Newark",
  "Wooster",
] as const;

export const CampusSchema: ZodSchema = z.enum(CampusTuple);
export type Campus = z.infer<typeof CampusSchema>;
