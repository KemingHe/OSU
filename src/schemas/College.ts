// ./src/schemas/College.ts
//
// OSU College ZodSchema and type definition,
// plus corresponding native TypeScript tuple definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

export const CollegeTuple = [
  "Architecture",
  "Arts and Sciences",
  "Business",
  "Dental Hygiene",
  "Dentistry",
  "Education and Human Ecology",
  "Engineering",
  "Environment and Natural Resources",
  "Food, Agricultural and Environmental Sciences",
  "Health and Rehabilitation Sciences",
  "Medicine",
  "Nursing",
  "Optometry",
  "Pharmacy",
  "Pre-Professional",
  "Public Affairs",
  "Public Health",
  "Social Work",
  "University Exploration",
] as const;

export const CollegeSchema:ZodSchema = z.enum(CollegeTuple);
export type College = z.infer<typeof CollegeSchema>;