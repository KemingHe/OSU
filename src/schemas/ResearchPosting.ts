// ./src/schemas/ResearchPosting.ts
//
// ResearchPosting ZodSchema and type definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

export const ResearchPostingSchema:ZodSchema = z.object({
  title              : z.string().min(1),
  link               : z.string().url(),
  applicationDeadline: z.nullable(z.string().min(1)),
  department         : z.nullable(z.string().min(1)),
  publicOrPrivate    : z.nullable(z.string().min(1)),
  hoursPerWeek       : z.nullable(z.string().min(1)),
  compensationTypes  : z.nullable(z.array(z.string().min(1)).nonempty()),
});

/**
 * TypeScript type inferred from the `ResearchPosting` Zod schema.
 *
 * This type represents the structure of a research posting object as defined by the `ResearchPosting` schema.
 *
 * @typedef {Object} ResearchPosting
 * @property {string} title - Title of the research posting, represented by a non-empty string.
 * @property {string} link - URL linking to the research posting, represented by a valid URL string.
 * @property {string | null} applicationDeadline - Application deadline, represented by a non-empty string, or null if not applicable or missing data.
 * @property {string | null} department - Department offering the research posting, represented by a non-empty string, or null if not applicable or missing data.
 * @property {string | null} publicOrPrivate - Indicates whether the posting is public or private, represented by a non-empty string, or null if not applicable or missing data.
 * @property {string | null} hoursPerWeek - Number of hours per week required, represented by a non-empty string, or null if not applicable or missing data.
 * @property {string[] | null} compensationTypes - Types of compensation offered, represented by a non-empty array of non-empty strings, or null if not applicable or missing data.
 */
export type ResearchPosting = z.infer<typeof ResearchPostingSchema>;