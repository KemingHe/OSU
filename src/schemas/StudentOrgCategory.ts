// ./src/schemas/StudentOrgCategory.ts
//
// StudentOrgCategory ZodSchema and type definition,
// plus corresponding native TypeScript tuple definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

export const StudentOrgCategoryTuple = [
  "Academic/College",
  "Awareness/Activism",
  "Community Service/Service Learning",
  "Creative and Performing Arts",
  "Ethnic/Cultural",
  "Governance Organizations",
  "Honoraries/Honor Societies",
  "Media, Journalism, and Creative Writing",
  "Religious/Spiritual",
  "Social Fraternities/Sororities",
  "Special Interest",
  "Sports and Recreation",
  "Technology",
  "Graduate",
  "Professional",
  "Undergraduate",
] as const;

export const StudentOrgCategorySchema: ZodSchema = z.enum(StudentOrgCategoryTuple);
export type StudentOrgCategory = z.infer<typeof StudentOrgCategorySchema>;