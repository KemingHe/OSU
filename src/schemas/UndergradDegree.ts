// ./src/schemas/UndergradDegree.ts
//
// UndergradDegree ZodSchema and type definition,
// plus corresponding native TypeScript tuple definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

export const UndergradDegreeTuple = [
  "AA (Associate of Arts)",
  "AS (Associate of Science)",
  "BA (Bachelor of Arts)",
  "BS (Bachelor of Science)",
  "BFA (Bachelor of Fine Arts)",
  "BM (Bachelor of Music)",
  "BME (Bachelor of Music Education)",
  "BSD (Bachelor of Science in Design)",
] as const;

export const UndergradDegreeSchema: ZodSchema = z.enum(UndergradDegreeTuple);
export type UndergradDegree = z.infer<typeof UndergradDegreeSchema>;