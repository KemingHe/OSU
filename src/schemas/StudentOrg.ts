// ./src/schemas/StudentOrg.ts
//
// StudentOrg ZodSchema and type definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

// Local ZodSchema imports.
import { CampusSchema } from "@src/schemas/Campus";
import { StudentOrgCategorySchema } from "@src/schemas/StudentOrgCategory";

export const StudentOrgSchema: ZodSchema = z.object({
  name            : z.string().min(1),
  purposeStatement: z.nullable(z.string().min(1)),
  campuses        : z.nullable(z.array(CampusSchema).nonempty()),
  categories      : z.nullable(z.array(StudentOrgCategorySchema).nonempty()),
});

/**
 * TypeScript type inferred from the `StudentOrg` Zod schema.
 *
 * This type represents the structure of a student organization object as defined by the `StudentOrg` schema.
 *
 * @typedef {Object} StudentOrg
 * @property {string} name - Name of the student organization, represented by a non-empty string.
 * @property {string | null} purposeStatement - Purpose statement of the student organization, represented by a non-empty string, or null if not applicable or missing data.
 * @property {Campus[] | null} campuses - Campuses where the student organization is active, represented by a non-empty array of `Campus` objects, or null if not applicable or missing data.
 * @property {StudentOrgCategory[] | null} categories - Categories of the student organization, represented by a non-empty array of `StudentOrgCategory` objects, or null if not applicable or missing data.
 */
export type StudentOrg = z.infer<typeof StudentOrgSchema>;
