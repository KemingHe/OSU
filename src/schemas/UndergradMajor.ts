// ./src/schemas/UndergradMajor.ts
//
// UndergradMajor ZodSchema and type definition.

// Zod essential imports.
import { z, type ZodSchema } from "zod";

// Local ZodSchema imports.
import { CampusSchema } from "@src/schemas/Campus";
import { CollegeSchema } from "@src/schemas/College";
import { UndergradDegreeSchema } from "@src/schemas/UndergradDegree";

export const UndergradMajorSchema: ZodSchema = z.object({
  major   : z.string().min(1),
  degrees : z.nullable(z.array(UndergradDegreeSchema).nonempty()),
  campuses: z.nullable(z.array(CampusSchema).nonempty()),
  college : z.nullable(CollegeSchema),
});

/**
 * TypeScript type inferred from the `UndergradMajor` Zod schema.
 *
 * This type represents the structure of an undergraduate major object as defined by the `UndergradMajor` schema.
 *
 * @typedef {Object} UndergradMajor
 * @property {string} major - Name of the major, represented by a non-empty string.
 * @property {UndergradDegree[] | null} degrees - Array of undergraduate degrees associated with the major, represented by a non-empty array of `UndergradDegree` objects, or null if not applicable or missing data.
 * @property {Campus[] | null} campuses - Campuses where the major is offered, represented by a non-empty array of `Campus` objects, or null if not applicable or missing data.
 * @property {College | null} college - College where the major belongs, represented by a `College` object, or null if not applicable or missing data.
 */
export type UndergradMajor = z.infer<typeof UndergradMajorSchema>;