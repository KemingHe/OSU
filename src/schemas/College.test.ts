// ./src/schemas/College.test.ts
//
// Unittests for the OSU college tuple and zod schema.

// Vitest essential imports.
import { describe, expect, it } from "vitest";

// Testing module imports.
import { CollegeTuple, CollegeSchema } from "@src/schemas/College";

// -----------------------------------------------------------------------------
// Constant valid and invalid college strings for batch testing.
const validCollegeStrings: string[] = [
  "Architecture",
  "Arts and Sciences",
  "Business",
  "Dental Hygiene",
  "Dentistry",
  "Education and Human Ecology",
  "Engineering",
  "Environment and Natural Resources",
  "Food, Agricultural and Environmental Sciences",
  "Health and Rehabilitation Sciences",
  "Medicine",
  "Nursing",
  "Optometry",
  "Pharmacy",
  "Pre-Professional",
  "Public Affairs",
  "Public Health",
  "Social Work",
  "University Exploration",
];

const invalidCollegeStrings: string[] = [
  "",
  "@",
  "123",
  "Nursing2",
  "Architecture ",
  "  Arts and Sciences  ",
  "College That Doesn't Exist",
];

// -----------------------------------------------------------------------------
// Test suite for the CollegeTuple constant.
describe("The CollegeTuple constant", () => {
  it("matches the snapshot", () => {
    expect(CollegeTuple).toMatchSnapshot();
  });
});

// -----------------------------------------------------------------------------
// Test suite for the CollegeSchema zod schema.
describe("The CollegeSchema zod schema", () => {
  it("matches the snapshot", () => {
    expect(CollegeSchema).toMatchSnapshot();
  });

  it("parses valid college strings", () => {
    for (const validCollegeString of validCollegeStrings) {
      expect(CollegeSchema.safeParse(validCollegeString).success).toBe(true);
    }
  });

  it("fails to parse invalid college strings", () => {
    for (const invalidCollegeString of invalidCollegeStrings) {
      expect(CollegeSchema.safeParse(invalidCollegeString).success).toBe(false);
    }
  });
});
