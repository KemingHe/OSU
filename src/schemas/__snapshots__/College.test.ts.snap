// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The CollegeSchema zod schema > matches the snapshot 1`] = `
ZodEnum {
  "_def": {
    "typeName": "ZodEnum",
    "values": [
      "Architecture",
      "Arts and Sciences",
      "Business",
      "Dental Hygiene",
      "Denistry",
      "Education and Human Ecology",
      "Engineering",
      "Environment and Natural Resources",
      "Food, Agricultural and Environmental Sciences",
      "Health and Rehabilitation Sciences",
      "Medicine",
      "Nursing",
      "Optometry",
      "Pharmacy",
      "Pre-Professional",
      "Public Affairs",
      "Public Health",
      "Social Work",
      "University Exploration",
    ],
  },
  "and": [Function],
  "array": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
}
`;

exports[`The CollegeTuple constant > matches the snapshot 1`] = `
[
  "Architecture",
  "Arts and Sciences",
  "Business",
  "Dental Hygiene",
  "Denistry",
  "Education and Human Ecology",
  "Engineering",
  "Environment and Natural Resources",
  "Food, Agricultural and Environmental Sciences",
  "Health and Rehabilitation Sciences",
  "Medicine",
  "Nursing",
  "Optometry",
  "Pharmacy",
  "Pre-Professional",
  "Public Affairs",
  "Public Health",
  "Social Work",
  "University Exploration",
]
`;
